// @ngageoint/geopackage-geojson-js/built/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GeoJSONToGeoPackage = void 0;
var geopackage_1 = require("@ngageoint/geopackage");
var fs_1 = __importDefault(require("fs"));
var path_1 = __importDefault(require("path"));
var bbox_1 = __importDefault(require("@turf/bbox"));
if (typeof window === 'undefined') {
    geopackage_1.setCanvasKitWasmLocateFile(function (file) {
        return path_1.default.join(__dirname, file);
    });
}
var GeoJSONToGeoPackage = /** @class */ (function () {
    function GeoJSONToGeoPackage(options) {
        this.options = options;
    }
    GeoJSONToGeoPackage.prototype._calculateTrueExtentForFeatureTable = function (gp, tableName) {
        var extent = undefined;
        var featureDao = gp.getFeatureDao(tableName);
        if (featureDao.isIndexed()) {
            if (featureDao.featureTableIndex.rtreeIndexDao != null) {
                var iterator = featureDao.featureTableIndex.rtreeIndexDao.queryForEach();
                var nextRow = iterator.next();
                while (!nextRow.done) {
                    if (extent == null) {
                        extent = [nextRow.value.minx, nextRow.value.miny, nextRow.value.maxx, nextRow.value.maxy];
                    }
                    else {
                        extent[0] = Math.min(extent[0], nextRow.value.minx);
                        extent[1] = Math.min(extent[1], nextRow.value.miny);
                        extent[2] = Math.max(extent[2], nextRow.value.maxx);
                        extent[3] = Math.max(extent[3], nextRow.value.maxy);
                    }
                    nextRow = iterator.next();
                }
            }
            else if (featureDao.featureTableIndex.geometryIndexDao != null) {
                var iterator = featureDao.featureTableIndex.geometryIndexDao.queryForEach();
                var nextRow = iterator.next();
                while (!nextRow.done) {
                    if (extent == null) {
                        extent = [nextRow.value.min_x, nextRow.value.min_y, nextRow.value.max_x, nextRow.value.max_y];
                    }
                    else {
                        extent[0] = Math.min(extent[0], nextRow.value.min_x);
                        extent[1] = Math.min(extent[1], nextRow.value.min_y);
                        extent[2] = Math.max(extent[2], nextRow.value.max_x);
                        extent[3] = Math.max(extent[3], nextRow.value.max_y);
                    }
                    nextRow = iterator.next();
                }
            }
        }
        if (extent == null) {
            var iterator = featureDao.queryForEach();
            var nextRow = iterator.next();
            while (!nextRow.done) {
                var featureRow = featureDao.getRow(nextRow.value);
                if (featureRow.geometry != null && featureRow.geometry.envelope != null) {
                    if (extent == null) {
                        extent = [
                            featureRow.geometry.envelope.minX,
                            featureRow.geometry.envelope.minY,
                            featureRow.geometry.envelope.maxX,
                            featureRow.geometry.envelope.maxY,
                        ];
                    }
                    else {
                        extent[0] = Math.min(extent[0], featureRow.geometry.envelope.minX);
                        extent[1] = Math.min(extent[1], featureRow.geometry.envelope.minY);
                        extent[2] = Math.max(extent[2], featureRow.geometry.envelope.maxX);
                        extent[3] = Math.max(extent[3], featureRow.geometry.envelope.maxY);
                    }
                }
                nextRow = iterator.next();
            }
        }
        return extent;
    };
    GeoJSONToGeoPackage.prototype._updateBoundingBoxForFeatureTable = function (gp, tableName) {
        var contentsDao = gp.contentsDao;
        var contents = contentsDao.queryForId(tableName);
        var extent = this._calculateTrueExtentForFeatureTable(gp, tableName);
        if (extent != null) {
            contents.min_x = extent[0];
            contents.min_y = extent[1];
            contents.max_x = extent[2];
            contents.max_y = extent[3];
        }
        else {
            contents.min_x = -180.0;
            contents.min_y = -90.0;
            contents.max_x = 180.0;
            contents.max_y = 90.0;
        }
        contentsDao.update(contents);
    };
    GeoJSONToGeoPackage.prototype.addLayer = function (options, progressCallback) {
        return __awaiter(this, void 0, void 0, function () {
            var clonedOptions;
            return __generator(this, function (_a) {
                clonedOptions = __assign(__assign({}, this.options), options);
                clonedOptions.append = true;
                return [2 /*return*/, this.setupConversion(clonedOptions, progressCallback)];
            });
        });
    };
    GeoJSONToGeoPackage.prototype.convert = function (options, progressCallback) {
        return __awaiter(this, void 0, void 0, function () {
            var clonedOptions;
            return __generator(this, function (_a) {
                clonedOptions = __assign(__assign({}, this.options), options);
                clonedOptions.append = false;
                return [2 /*return*/, this.setupConversion(clonedOptions, progressCallback)];
            });
        });
    };
    GeoJSONToGeoPackage.prototype.extract = function (geopackage, tableName) {
        return __awaiter(this, void 0, void 0, function () {
            var geoJson, iterator, iterator_1, iterator_1_1, feature;
            var e_1, _a;
            return __generator(this, function (_b) {
                geoJson = {
                    type: 'FeatureCollection',
                    features: [],
                };
                iterator = geopackage.iterateGeoJSONFeatures(tableName);
                try {
                    for (iterator_1 = __values(iterator), iterator_1_1 = iterator_1.next(); !iterator_1_1.done; iterator_1_1 = iterator_1.next()) {
                        feature = iterator_1_1.value;
                        geoJson.features.push(feature);
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (iterator_1_1 && !iterator_1_1.done && (_a = iterator_1.return)) _a.call(iterator_1);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
                return [2 /*return*/, Promise.resolve(geoJson)];
            });
        });
    };
    GeoJSONToGeoPackage.prototype.setupConversion = function (options, progressCallback) {
        return __awaiter(this, void 0, void 0, function () {
            var geoPackage, srsNumber, geoJson, tableName, name, nameSuffix, tables, count, featureCollection, properties, i, feature;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        geoPackage = options.geoPackage;
                        srsNumber = options.srsNumber || 4326;
                        geoJson = options.geoJson;
                        tableName = options.tableName;
                        return [4 /*yield*/, this.createOrOpenGeoPackage(geoPackage, options, progressCallback)];
                    case 1:
                        geoPackage = _a.sent();
                        if (typeof geoJson === 'string') {
                            name = path_1.default.basename(geoJson, path_1.default.extname(geoJson));
                        }
                        name = tableName || name || 'features';
                        nameSuffix = '';
                        tables = geoPackage.getFeatureTables();
                        count = 1;
                        while (tables.indexOf(name + nameSuffix) !== -1) {
                            nameSuffix = '_' + count;
                            count++;
                        }
                        tableName = name + nameSuffix;
                        if (!(typeof geoJson === 'string')) return [3 /*break*/, 5];
                        if (!progressCallback) return [3 /*break*/, 3];
                        return [4 /*yield*/, progressCallback({ status: 'Reading GeoJSON file' })];
                    case 2:
                        _a.sent();
                        _a.label = 3;
                    case 3: return [4 /*yield*/, new Promise(function (resolve, reject) {
                            fs_1.default.readFile(geoJson, 'utf8', function (err, data) {
                                resolve(JSON.parse(data));
                            });
                        })];
                    case 4:
                        geoJson = _a.sent();
                        _a.label = 5;
                    case 5:
                        featureCollection = {
                            type: 'FeatureCollection',
                            features: [],
                        };
                        properties = {};
                        for (i = 0; i < geoJson.features.length; i++) {
                            feature = geoJson.features[i];
                            this.addFeatureProperties(feature, properties);
                            if (feature.geometry !== null) {
                                featureCollection.features.push(feature);
                            }
                            else {
                                featureCollection.features.push({
                                    type: 'Feature',
                                    properties: feature.properties,
                                    geometry: null,
                                });
                            }
                        }
                        return [2 /*return*/, this.convertGeoJSONToGeoPackage(featureCollection, geoPackage, tableName, properties, srsNumber, progressCallback)];
                }
            });
        });
    };
    GeoJSONToGeoPackage.prototype.addFeatureProperties = function (feature, currentProperties) {
        if (feature.properties.geometry) {
            feature.properties.geometry_property = feature.properties.geometry;
            delete feature.properties.geometry;
        }
        if (feature.id) {
            if (!currentProperties['_feature_id']) {
                currentProperties['_feature_id'] = currentProperties['_feature_id'] || {
                    name: '_feature_id',
                    type: 'DOUBLE',
                };
            }
        }
        for (var key in feature.properties) {
            if (!currentProperties[key]) {
                var type = typeof feature.properties[key];
                if (feature.properties[key] !== undefined && feature.properties[key] !== null && type !== 'undefined') {
                    if (type === 'object') {
                        if (feature.properties[key] instanceof Date) {
                            type = 'Date';
                        }
                        else {
                            continue;
                        }
                    }
                    switch (type) {
                        case 'Date':
                            type = 'DATETIME';
                            break;
                        case 'number':
                            type = 'DOUBLE';
                            break;
                        case 'string':
                            type = 'TEXT';
                            break;
                        case 'boolean':
                            type = 'BOOLEAN';
                            break;
                    }
                    currentProperties[key] = {
                        name: key,
                        type: type,
                    };
                }
            }
        }
    };
    GeoJSONToGeoPackage.prototype.convertGeoJSONToGeoPackage = function (geoJson, geoPackage, tableName, properties, srsNumber, progressCallback) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.convertGeoJSONToGeoPackageWithSrs(geoJson, geoPackage, tableName, properties, srsNumber, progressCallback)];
            });
        });
    };
    GeoJSONToGeoPackage.prototype.convertGeoJSONToGeoPackageWithSrs = function (geoJson, geoPackage, tableName, properties, srsNumber, progressCallback) {
        return __awaiter(this, void 0, void 0, function () {
            var geometryColumns, columns, index, key, prop, tmp, boundingBox, featureDao, srs, count, featureCount, fivePercent, i, feature;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        geometryColumns = new geopackage_1.GeometryColumns();
                        geometryColumns.table_name = tableName;
                        geometryColumns.column_name = 'geometry';
                        geometryColumns.geometry_type_name = 'GEOMETRY';
                        geometryColumns.z = 2;
                        geometryColumns.m = 2;
                        columns = [];
                        columns.push(geopackage_1.FeatureColumn.createPrimaryKeyColumn(0, 'id', true));
                        columns.push(geopackage_1.FeatureColumn.createGeometryColumn(1, 'geometry', geopackage_1.GeometryType.GEOMETRY, false, null));
                        index = 2;
                        for (key in properties) {
                            prop = properties[key];
                            if (prop.name.toLowerCase() !== 'id') {
                                columns.push(geopackage_1.FeatureColumn.createColumn(index, prop.name, geopackage_1.GeoPackageDataType.fromName(prop.type)));
                            }
                            else {
                                columns.push(geopackage_1.FeatureColumn.createColumn(index, '_properties_' + prop.name, geopackage_1.GeoPackageDataType.fromName(prop.type), false, null));
                            }
                            index++;
                        }
                        if (!progressCallback) return [3 /*break*/, 2];
                        return [4 /*yield*/, progressCallback({ status: 'Creating table "' + tableName + '"' })];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2:
                        tmp = bbox_1.default(geoJson);
                        boundingBox = new geopackage_1.BoundingBox(Math.max(-180, tmp[0]), Math.min(180, tmp[2]), Math.max(-90, tmp[1]), Math.min(90, tmp[3]));
                        if (geoPackage
                            .getFeatureTables()
                            .map(function (table) { return table.toLowerCase(); })
                            .indexOf(tableName.toLowerCase()) === -1) {
                            console.log('creating feature table: ' + tableName);
                            geoPackage.createFeatureTable(tableName, geometryColumns, columns, boundingBox, srsNumber);
                        }
                        featureDao = geoPackage.getFeatureDao(tableName);
                        srs = featureDao.srs;
                        count = 0;
                        featureCount = geoJson.features.length;
                        fivePercent = Math.floor(featureCount / 20);
                        i = 0;
                        _a.label = 3;
                    case 3:
                        if (!(i < featureCount)) return [3 /*break*/, 6];
                        feature = geoJson.features[i];
                        if (feature.id) {
                            feature.properties._feature_id = feature.id;
                        }
                        if (feature.properties.id) {
                            feature.properties._properties_id = feature.properties.id;
                            delete feature.properties.id;
                        }
                        if (feature.properties.ID) {
                            feature.properties._properties_ID = feature.properties.ID;
                            delete feature.properties.ID;
                        }
                        geoPackage.addGeoJSONFeatureToGeoPackageWithFeatureDaoAndSrs(feature, featureDao, srs, true);
                        if (!(count++ % fivePercent === 0)) return [3 /*break*/, 5];
                        if (!progressCallback) return [3 /*break*/, 5];
                        return [4 /*yield*/, progressCallback({
                                status: 'Inserting features into table "' + tableName + '"',
                                completed: count,
                                total: featureCount,
                            })];
                    case 4:
                        _a.sent();
                        _a.label = 5;
                    case 5:
                        i++;
                        return [3 /*break*/, 3];
                    case 6:
                        if (!progressCallback) return [3 /*break*/, 8];
                        return [4 /*yield*/, progressCallback({
                                status: 'Done inserting features into table "' + tableName + '"',
                            })];
                    case 7:
                        _a.sent();
                        _a.label = 8;
                    case 8:
                        this._updateBoundingBoxForFeatureTable(geoPackage, tableName);
                        return [2 /*return*/, geoPackage];
                }
            });
        });
    };
    GeoJSONToGeoPackage.prototype.createOrOpenGeoPackage = function (geoPackage, options, progressCallback) {
        return __awaiter(this, void 0, void 0, function () {
            var stats;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!(typeof geoPackage === 'object')) return [3 /*break*/, 3];
                        if (!progressCallback) return [3 /*break*/, 2];
                        return [4 /*yield*/, progressCallback({ status: 'Opening GeoPackage' })];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2: return [2 /*return*/, geoPackage];
                    case 3:
                        stats = void 0;
                        try {
                            stats = fs_1.default.statSync(geoPackage);
                        }
                        catch (e) { }
                        if (stats && !options.append) {
                            throw new Error('GeoPackage file already exists, refusing to overwrite ' + geoPackage);
                        }
                        else if (stats) {
                            return [2 /*return*/, geopackage_1.GeoPackageAPI.open(geoPackage)];
                        }
                        if (!progressCallback) return [3 /*break*/, 5];
                        return [4 /*yield*/, progressCallback({ status: 'Creating GeoPackage' })];
                    case 4:
                        _a.sent();
                        _a.label = 5;
                    case 5: return [2 /*return*/, geopackage_1.GeoPackageAPI.create(geoPackage)];
                }
            });
        });
    };
    return GeoJSONToGeoPackage;
}());
exports.GeoJSONToGeoPackage = GeoJSONToGeoPackage;
