// @ngageoint/projections-js/dist/lib/Projection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Projection = void 0;
var internal_1 = require("./internal");
/**
 * Single Projection for an authority and code
 */
var Projection = /** @class */ (function () {
    /**
     * Constructor
     *
     * @param authority
     *            coordinate authority
     * @param code
     *            coordinate code
     * @param definition
     *            proj4 definition
     */
    function Projection(authority, code, definition) {
        this.authority = authority;
        this.code = code;
        this.definition = definition;
    }
    /**
     * Get the coordinate authority
     *
     * @return authority
     */
    Projection.prototype.getAuthority = function () {
        return this.authority;
    };
    /**
     * Get the coordinate code
     *
     * @return code
     */
    Projection.prototype.getCode = function () {
        return this.code;
    };
    /**
     * Get the projection definition
     * @return definition
     */
    Projection.prototype.getDefinition = function () {
        return this.definition;
    };
    /**
     * Get the transformation from this Projection to the EPSG code. Each thread
     * of execution should have it's own transformation.
     *
     * @param projection projection
     * @return transform
     */
    Projection.prototype.getTransformation = function (projection) {
        return internal_1.Projections.getProjectionTransformation(this, projection);
    };
    /**
     * Get the inverse transformation from this Projection to the EPSG code. Each thread
     * of execution should have it's own transformation.
     *
     * @param projection projection
     * @return transform
     */
    Projection.prototype.getInverseTransformation = function (projection) {
        return internal_1.Projections.getProjectionTransformation(projection, this);
    };
    /**
     * Get the transformation from this Projection to the EPSG code. Each thread
     * of execution should have it's own transformation.
     *
     *
     * @param authority
     *            coordinate authority
     * @param code
     *            coordinate code
     * @return transform
     */
    Projection.prototype.getTransformationTo = function (authority, code) {
        if (authority === void 0) { authority = internal_1.ProjectionConstants.AUTHORITY_EPSG; }
        return internal_1.Projections.getProjectionTransformation(this, internal_1.Projections.getProjection(authority, code));
    };
    /**
     * Get the inverse transformation from this Projection to the EPSG code. Each thread
     * of execution should have it's own transformation.
     *
     *
     * @param authority
     *            coordinate authority
     * @param code
     *            coordinate code
     * @return transform
     */
    Projection.prototype.getInverseTransformationTo = function (authority, code) {
        if (authority === void 0) { authority = internal_1.ProjectionConstants.AUTHORITY_EPSG; }
        return internal_1.Projections.getProjectionTransformation(this, internal_1.Projections.getProjection(authority, code));
    };
    /**
     * Check if this projection is equal to the authority and code
     * @param authority coordinate authority
     * @param code coordinate code
     * @return true if equal
     */
    Projection.prototype.equals = function (authority, code) {
        return this.authority === authority && this.code === code;
    };
    /**
     * Check if this projection is equal to the authority and code
     * @param authority coordinate authority
     * @param code coordinate code
     * @return true if equal
     */
    Projection.prototype.equalsProjection = function (other) {
        return this.authority === other.authority && this.code === other.code;
    };
    /**
     * {@inheritDoc}
     */
    Projection.prototype.toString = function () {
        return this.authority + ":" + this.code;
    };
    return Projection;
}());
exports.Projection = Projection;
