// @ngageoint/projections-js/dist/lib/ProjectionTransform.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectionTransform = void 0;
var proj4_1 = __importDefault(require("proj4"));
var internal_1 = require("./internal");
var ProjectionTransform = /** @class */ (function () {
    function ProjectionTransform(from, to) {
        this.from = from;
        this.to = to;
        // @ts-ignore
        this.converter = (0, proj4_1.default)(from.getDefinition(), to.getDefinition());
        if (internal_1.Projections.projectionsEqual(from, to)) {
            this._transform = function (x, y) {
                return [x, y];
            };
        }
        else if (internal_1.Projections.isWGS84(from) && internal_1.Projections.isWebMercator(to)) {
            this._transform = function (x, y) {
                return this.converter.forward([Math.max(-internal_1.ProjectionConstants.WGS84_HALF_WORLD_LON_WIDTH, Math.min(internal_1.ProjectionConstants.WGS84_HALF_WORLD_LON_WIDTH, x)), Math.max(internal_1.ProjectionConstants.WEB_MERCATOR_MIN_LAT_RANGE, Math.min(internal_1.ProjectionConstants.WEB_MERCATOR_MAX_LAT_RANGE, y))]);
            };
        }
        else {
            this._transform = function (x, y) {
                return this.converter.forward([x, y]);
            };
        }
    }
    /**
     * Transform a x and y location
     * @param x x coordinate
     * @param y y coordinate
     * @return transformed coordinates as [x, y]
     */
    ProjectionTransform.prototype.transform = function (x, y) {
        return this._transform(x, y);
    };
    /**
     * Transform a coordinates array
     * @param array coordinate
     * @return transformed coordinates as [x, y]
     */
    ProjectionTransform.prototype.transformCoordinateArray = function (array) {
        return this._transform(array[0], array[1]);
    };
    /**
     * Transform InterfaceCoordinates
     * @return transformed coordinates as {x, y}
     * @param coordinate
     */
    ProjectionTransform.prototype.transformCoordinate = function (coordinate) {
        var bounds = this._transform(coordinate.x, coordinate.y);
        return {
            x: bounds[0],
            y: bounds[1]
        };
    };
    /**
     * Transform the coordinate bounds
     *
     * @param minX min x
     * @param minY min y
     * @param maxX max x
     * @param maxY max y
     * @return transformed coordinate bounds as [minX, minY, maxX, maxY]
     */
    ProjectionTransform.prototype.transformBounds = function (minX, minY, maxX, maxY) {
        var projectedLowerLeft = this.transform(minX, minY);
        var projectedLowerRight = this.transform(maxX, minY);
        var projectedUpperRight = this.transform(maxX, maxY);
        var projectedUpperLeft = this.transform(minX, maxY);
        var bounds = [];
        bounds[0] = Math.min(projectedLowerLeft[0], projectedLowerRight[0], projectedUpperRight[0], projectedUpperLeft[0]);
        bounds[1] = Math.min(projectedLowerLeft[1], projectedLowerRight[1], projectedUpperRight[1], projectedUpperLeft[1]);
        bounds[2] = Math.max(projectedLowerLeft[0], projectedLowerRight[0], projectedUpperRight[0], projectedUpperLeft[0]);
        bounds[3] = Math.max(projectedLowerLeft[1], projectedLowerRight[1], projectedUpperRight[1], projectedUpperLeft[1]);
        return bounds;
    };
    /**
     * Get the from projection in the transform
     * @return from projection
     */
    ProjectionTransform.prototype.getFromProjection = function () {
        return this.from;
    };
    /**
     * Get the to projection in the transform
     * @return to projection
     */
    ProjectionTransform.prototype.getToProjection = function () {
        return this.to;
    };
    /**
     * Get the transform
     * @return transform
     */
    ProjectionTransform.prototype.getConverter = function () {
        return this.converter;
    };
    ProjectionTransform.prototype.getInverseTransformation = function () {
        return new ProjectionTransform(this.to, this.from);
    };
    return ProjectionTransform;
}());
exports.ProjectionTransform = ProjectionTransform;
